### PR TITLE
SslStream WriteAsync

### PR DESCRIPTION
**Pull request by benaadams**
This is just the WriteAsync portion of https://github.com/dotnet/corefx/issues/5077

The ReadAsync looks a lot more complicated so I'm not even going to attempt (i.e. the Read wasn't immediately apparent to me)

### CODE REVIEW

**Comment by davidsh** [comment]
cc: @CIPop @stephentoub

**Comment by benaadams** [comment]
Actually ReadAsync doesn't look too bad, other than the effects of async all the way down...

File: src/System.Net.Security/src/System/Net/SecureProtocols/SslStreamInternal.cs [hunk 1]
```diff
@@ -229,6 +235,80 @@ private void ProcessWrite(byte[] buffer, int offset, int count)
...
         }
 
+        private async Task ProcessWriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            ValidateParameters(buffer, offset, count);
+
+            if (Interlocked.Exchange(ref _NestedWrite, 1) == 1)
+            {
+                throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, "Write", "write"));
+            }
```
Comment (Maxwe11) [inline src/System.Net.Security/src/System/Net/SecureProtocols/SslStreamInternal.cs]: I'd rather move this validation out of `async` scope.

File: src/System.Net.Security/src/System/Net/SecureProtocols/SslStreamInternal.cs [hunk 2]
```diff
@@ -229,6 +235,80 @@ private void ProcessWrite(byte[] buffer, int offset, int count)
...
+
+            try
+            {
+                await StartWritingAsync(buffer, offset, count, cancellationToken);
```
Comment (Maxwe11) [inline src/System.Net.Security/src/System/Net/SecureProtocols/SslStreamInternal.cs]: `ConfigureAwait(false)` ?

File: src/System.Net.Security/src/System/Net/SecureProtocols/SslStreamInternal.cs (hunk 1 above)
Comment (benaadams) [inline src/System.Net.Security/src/System/Net/SecureProtocols/SslStreamInternal.cs]: Done. Also have moved it up for sync so they remain matching implementations

File: src/System.Net.Security/src/System/Net/SecureProtocols/SslStreamInternal.cs [hunk 3]
```diff
@@ -268,21 +369,9 @@ private void StartWriting(byte[] buffer, int offset, int count)
...
+                    int encryptedBytes = EncryptWritingBuffer(buffer, offset, outBuffer, chunkBytes);
 
-                    if (PinnableBufferCacheEventSource.Log.IsEnabled())
-                    {
-                        PinnableBufferCacheEventSource.Log.DebugMessage3("In System.Net._SslStream.StartWriting Got Encrypted Buffer",
-                            this.GetHashCode(), encryptedBytes, PinnableBufferCacheEventSource.AddressOfByteArray(outBuffer));
-                    }
-
-                    _SslState.InnerStream.Write(outBuffer, 0, encryptedBytes);
+                    await _SslState.InnerStream.WriteAsync(outBuffer, 0, encryptedBytes, cancellationToken);
```
Comment (benaadams) [inline src/System.Net.Security/src/System/Net/SecureProtocols/SslStreamInternal.cs]: ConfigureAwait(false)

**Comment by benaadams** [comment]
Going through Read in more detail; which really is more complicated, and since its the security stream I don't think its advisable for me to try to convert it. So will just leave my contribution as the WriteAsync portion and bow out...

**Comment by CIPop** [comment]
Thanks @benaadams. Please add tests to validate the new behaviors.
Also, these tests should pass in both .Net Desktop and CoreFX to ensure app-compat.

**Comment by CIPop** [comment]
> The ReadAsync looks a lot more complicated so I'm not even going to attempt

@benaadams Given the code complexity of the original .Net Desktop code, I'd rather retry to port the code from [_SslStream.cs](http://referencesource.microsoft.com/#System/net/System/Net/SecureProtocols/_SslStream.cs,9c5cd436814e7aa7) instead of creating new TPL APIs (i.e. merging this PR). After the Begin/End APIs are implemented, we can add TPL overrides to Read/WriteAsync based on wrappers similar to the ones in NetworkStream and NegotiateStream.

**Comment by benaadams** [comment]
@CIPop I'm good with that

**Pull request closed without merging**